contrib/pg_tde/documentation/docs/architecture.md

```diff
@@ -0,0 +1,403 @@
+
+### WAL encryption
+
+WAL encryption is currently controlled globally, it's either turned on or off with a global GUC variable that requires a server restart.
+The variable only controls writes, when it is turned on, WAL writes are encrypted.
+
+This means WAL files can contain both encrypted and unencrpyted data, depending on what the status of this variable was when writing the data.
+`pg_tde` keeps track of the encryption status of WAL records using internal keys:
+every time the encryption status of WAL changes, it writes a new internal key for WAL.
+When the encryption is turned on, this internal key contains a valid encryption key.
+When the encrpytion is turned off, it only contains a flag signaling that WAL encryption ended.
+
+With this information, the WAL reader code can decide if a specific WAL records has to be decrypted or not.
+
+### Principal key rotation
+
+To comply with common policies, and to handle situations with potentially exposed principal keys, principal keys can be rotated.
+Rotation means that `pg_tde` generates a new version of the principal key, and re-encrypts the associated internal keys with the new key.
+The old key is kept as is at the same location, as it is potentially still needed to decrypt backups or other databases.
+
+It is also possible that the location of the keyring changes:
+either the service is moved to a new address, or the keyring has to be moved to a different keyring provider type. Both of these situations are supported by `pg_tde` using simple SQL functions.
+
+In certain cases the SQL API can't be used for these operations:
+if the server isn't running, and the old keyring is no longer available, startup can fail if it needs to access the encryption keys. 
+For these situations, `pg_tde` also provides command line tools to recover the database.
+
+
+### Internal key regeneration
+
+Internal keys for files (tables, indexes, etc.) are fixed once the file is created, there's no way currently to re-encrypt a file.
+
+There are workarounds for this, because operations that move the table data to a new file, such as `VACUUM FULL` or an `ALTER TABLE` that rewrites the file will create a new key for the new file, essentially rotating the internal key.
+This however means taking an exclusive lock on the table for the duration of the operation, which might not be desirable for huge tables.
+
+### Internal key storage
+
+Internal keys, and generally `pg_tde` metadata is kept in a single directory in `$PGDATA/pg_tde`. 
+In this directory each database has separate files, containing:
+
+* Encrypted internal keys and internal key mapping to tables
+* Information about the key providers
+
+There's also a special global section marked with the OID 607, which includes the global key providers / global internal keys.
+
+This is used by the WAL encryption, and can optionally be used by specific databases too, if global provider inheritance is enabled.
```

Comment:
Don't they use the default provider for this purpose? 
If not, let's explain the default scope and default provider here too.